service/controller/resource/spark/create.go

```diff
@@ -0,0 +1,683 @@
+	var dataSecret *corev1.Secret
+	{
+		if sparkCR.Status.DataSecretName != "" {
+			r.logger.LogCtx(ctx, "level", "debug", "message", fmt.Sprintf("Spark CR status already references a Secret %#q", sparkCR.Status.DataSecretName))
+			var s corev1.Secret
+			err = r.ctrlClient.Get(ctx, client.ObjectKey{Namespace: azureMachinePool.Namespace, Name: sparkCR.Status.DataSecretName}, &s)
+			if errors.IsNotFound(err) {
+				// This is ok. We'll create it then.
+				r.logger.LogCtx(ctx, "level", "debug", "message", fmt.Sprintf("Secret %#q was not found, it will be created", sparkCR.Status.DataSecretName))
+			} else if err != nil {
+				return microerror.Mask(err)
+			} else {
+				dataSecret = &s
+			}
+		}
+
+		// Create Secret for ignition data.
+		if dataSecret == nil {
+			dataSecret = &corev1.Secret{
+				ObjectMeta: metav1.ObjectMeta{
+					Name:      secretName(sparkCR.Name),
+					Namespace: azureMachinePool.Namespace,
+					OwnerReferences: []metav1.OwnerReference{
+						{
+							APIVersion:         corev1alpha1.SchemeGroupVersion.String(),
+							BlockOwnerDeletion: to.BoolPtr(true),
+							Kind:               "Spark",
+							Name:               sparkCR.Name,
+							UID:                sparkCR.UID,
+						},
+					},
+				},
+				Data: map[string][]byte{
+					key.CloudConfigSecretKey: ignitionBlob,
+				},
+			}
+
+			err = r.ctrlClient.Create(ctx, dataSecret)
+			if err != nil {
+				return microerror.Mask(err)
+			}
+			r.logger.LogCtx(ctx, "level", "debug", "message", fmt.Sprintf("Secret %#q created with ignition cloud config", secretName(sparkCR.Name)))
+		}
+	}
+
+	{
+		if !bytes.Equal(ignitionBlob, dataSecret.Data[key.CloudConfigSecretKey]) {
+			r.logger.LogCtx(ctx, "level", "debug", "message", fmt.Sprintf("Ignition cloud config in Secret %#q is out of date, updating it", dataSecret.Name))
+
+			dataSecret.Data[key.CloudConfigSecretKey] = ignitionBlob
+
+			err = r.ctrlClient.Update(ctx, dataSecret)
+			if err != nil {
+				return microerror.Mask(err)
+			}
+		}
+	}
+
+	{
+		var sparkStatusUpdateNeeded bool
+
+		if !sparkCR.Status.Ready {
+			sparkCR.Status.Ready = true
+			sparkStatusUpdateNeeded = true
+		}
+
+		if sparkCR.Status.DataSecretName != dataSecret.Name {
+			sparkCR.Status.DataSecretName = dataSecret.Name
+			sparkStatusUpdateNeeded = true
+		}
+
+		if sparkCR.Status.Verification.Hash != dataHash {
+			sparkCR.Status.Verification.Hash = dataHash
+			sparkCR.Status.Verification.Algorithm = "sha512"
+			sparkStatusUpdateNeeded = true
+		}
+
+		if sparkStatusUpdateNeeded {
+			r.logger.LogCtx(ctx, "level", "debug", "message", fmt.Sprintf("Updating Spark CR %#q", sparkCR.Name))
+			err = r.ctrlClient.Status().Update(ctx, &sparkCR)
+			if err != nil {
+				return microerror.Mask(err)
+			}
+		}
+	}
+
+	return nil
+}
+
+func (r *Resource) createIgnitionBlob(ctx context.Context, azureMachinePool *expcapzv1alpha3.AzureMachinePool) ([]byte, error) {
+	machinePool, err := r.getOwnerMachinePool(ctx, azureMachinePool.ObjectMeta)
```

Comment:
This should work with our current owner references. Also, I'm now using the utility function to find the cluster by name.